lib/charms/prometheus_k8s/v0/prometheus.py

```diff
@@ -684,6 +794,77 @@ def _set_unit_ip(self, event):
                 self._charm.model.get_binding(relation).network.bind_address
             )
 
+    def _label_alert_topology(self, rule):
+        """Insert juju topology labels into an alert rule.
+
+        Args:
+            rule: a dictionary representing a Prometheus alert rule.
+
+        Returns:
+            a dictionary representing Prometheus alert rule with Juju
+            topology labels.
+        """
+        metadata = self._scrape_metadata
+        labels = rule.get("labels", {})
+        labels["juju_model"] = metadata["model"]
+        labels["juju_model_uuid"] = metadata["model_uuid"]
+        labels["juju_application"] = metadata["application"]
+        rule["labels"] = labels
+        return rule
+
+    def _label_alert_expression(self, rule):
+        """Insert juju topology filters into a Prometheus alert rule.
+
+        Args:
+            rule: a dictionary representing a Prometheus alert rule.
+
+        Returns:
+            a dictionary representing a Prometheus alert rule that filters based
+            on juju topology.
+        """
+        metadata = self._scrape_metadata
+        topology = (
+            'juju_model="{}", juju_model_uuid="{}", juju_application="{}"'.format(
+                metadata["model"], metadata["model_uuid"], metadata["application"]
+            )
+        )
+        expr = rule["expr"]  # a rule has to have an "expr"
+        expr = expr.replace("%%juju_topology%%", topology)
+        rule["expr"] = expr
+        return rule
```

Comment:
```suggestion
        if expr := rule.get("expr", None):  # a rule has to have an "expr"
            expr = expr.replace("%%juju_topology%%", topology)
            rule["expr"] = expr
            return rule
        # raise an exception?
```

Reply:
I am not convinced that `rule.get("expr", None)` is a good idea. Suppose a user creates an alert rule that does not have an alert expression (clearly and invalid alert rule), what should be the behaviour ? As it stands a `KeyError` exception will be raised. However with the proposed change the invalid rule will be silently ignored. I do not think being silent would be a good idea. I think it may be reasonable to throw and explicit custom exception here or set the charm into a blocked state. What would be better ?

Reply:
Yeh, so my comment at the end was an indication that there probably needs to be an else, or an exception raised to indicate that something invalid was specified

Reply:
The `if expr := rule.get("expr", None):` will evaluate false if a rule isnt set and give you the opportunity to handle that better

Reply:
OK. Will add a custom exception `InvalidAlertRule` implement, test, and document it then push. I copy and paste this suggested code change into my changes instead of creating commit within github (hope that is ok).

Reply:
of course :)

Reply:
This has been refactored and invalid expression now logs and error instead of raising a KeyError. Unit tests follow.